src/colmap/estimators/covariance.h

```diff
@@ -0,0 +1,174 @@
+                                double* paramter_block2) const;
+
+  // test if either "ComputeFull()" or "Compute()" has been called
+  bool HasValidPoseCovariance() const;
+  // test if "ComputeFull()" has been called
+  bool HasValidFullCovariance() const;
+
+ protected:
+  // blocks parsed from reconstruction (initialized at construction)
+  std::vector<const double*> pose_blocks_;
+  int num_params_poses_ = 0;
+  std::vector<const double*> other_variables_blocks_;
+  int num_params_other_variables_ = 0;
+  std::vector<const double*> point_blocks_;
+  int num_params_points_ = 0;
+
+  // get the starting index of the parameter block in the matrix
+  // orders: [pose_blocks, other_variables_blocks, point_blocks]
+  std::map<const double*, int> map_block_to_index_;
+  int GetBlockIndex(const double* params) const;
+  int GetBlockTangentSize(const double* params) const;
+  int GetPoseIndex(image_t image_id) const;
+  int GetPoseTangentSize(image_t image_id) const;
+
+  // covariance for all parameters (except for 3D points), equal to
+  // H_cc_.inverse()
+  Eigen::MatrixXd cov_variables_;
+
+  // covariance for pose parameters
+  Eigen::MatrixXd cov_poses_;
+
+  // ceres problem
+  ceres::Problem* problem_;
+
+  // reconstruction
+  Reconstruction* reconstruction_;
+
+ private:
+  // set up parameter blocks
+  void Setup();
+};
+
+class BundleAdjustmentCovarianceEstimatorCeresBackend
+    : public BundleAdjustmentCovarianceEstimatorBase {
+ public:
+  BundleAdjustmentCovarianceEstimatorCeresBackend(
+      ceres::Problem* problem, Reconstruction* reconstruction)
+      : BundleAdjustmentCovarianceEstimatorBase(problem, reconstruction) {}
+
+  bool ComputeFull() override;
+  bool Compute() override;
+};
+
+class BundleAdjustmentCovarianceEstimator
+    : public BundleAdjustmentCovarianceEstimatorBase {
+ public:
+  BundleAdjustmentCovarianceEstimator(ceres::Problem* problem,
+                                      Reconstruction* reconstruction)
+      : BundleAdjustmentCovarianceEstimatorBase(problem, reconstruction) {}
+
+  void SetLambda(double lambda);  // update daming factor. default = 1e-6
+
+  bool ComputeFull() override;
+  bool Compute() override;
+
+ private:
+  // The Schur complement for all parameters (except for 3D points) after Schur
+  // elimination
+  Eigen::SparseMatrix<double> S_matrix_;
+
+  // damping factor to avoid rank deficiency for poorly conditioned 3D points
+  double lambda_ = 1e-6;
+
+  // Compute the Schur complement for poses and other variables by eliminating
+  // 3D points
+  void ComputeSchurComplement();
+  bool HasValidSchurComplement() const;
+};
+
+bool EstimatePoseCovarianceCeresBackend(
+    ceres::Problem* problem,
+    Reconstruction* reconstruction,
+    std::map<image_t, Eigen::MatrixXd>& res);
+
+bool EstimatePoseCovariance(ceres::Problem* problem,
+                            Reconstruction* reconstruction,
+                            std::map<image_t, Eigen::MatrixXd>& res,
```

Comment:
We use MatrixXd to deal with constant rotation/translation components. Would be good to briefly document the ordering of the parameter blocks in the resulting matrix in the header here?

Reply:
> We use MatrixXd to deal with constant rotation/translation components. 

Do you mean non-constant?

Reply:
Yes, the matrix only contains the non constant parts, I assume?

Reply:
Yes. I will add some documentations in the header.